### DISCUSSION TITLE
File System

### DISCUSSION THREAD

**Discussion by itsamedood** [HaxeFlixel/flixel#2415] [Q&A] [answered]
I had an idea for my game to write to a text document with some secret stuff to scare the player, similar to DDLC, but I can't use the File system with HaxeFlixel :(

I know how to READ a file with `openfl.Assets.getText()`, but as far as I can tell there's no way to write to a file using `openfl.Assets`.

Any other way i can pull this off?

**Comment by Geokureli** [comment, marked as answer]
If you're making a desktop game, check out [sys.io.File](https://api.haxe.org/sys/io/File.html) and [sys.FileSystem](https://api.haxe.org/sys/FileSystem.html#FileSystem). but if you're making a web game, this isn't really feasible. 

For something like DDLC you would include the files in your assets folder in their initial state and make changes to them as the game progresses.

**Reply by itsamedood** [reply]
I don't really plan to do a web build, so that won't be an issue. I've already tried to use `sys.io.File` and `sys.FileSystem` but the game straight up crashes for me when attempting to do anything with it. My question is if there's another way, if I'm doing something wrong, or if I'm just screwed.

`Error:`

<img width="1019" alt="Screen Shot 2021-09-26 at 8 18 27 PM" src="https://user-images.githubusercontent.com/74739472/134829363-858d9448-349f-4b62-839f-8a305752486d.png">

**Reply by TheHENOOB** [reply]
The file you are trying to make the game read exists?

**Reply by itsamedood** [reply]
Yes. The path is correct and the file does in fact exist.

**Reply by Geokureli** [reply]
what path are you using?
Note in VSCode you can right click a flie and get the relative path

**Reply by itsamedood** [reply]
`assets/scripts/dialogue.txt`.

**Reply by Geokureli** [reply]
you might need to use [FileSystem.absolutePath(relativePath)](https://api.haxe.org/sys/FileSystem.html#absolutePath)

**Reply by itsamedood** [reply]
That's not exactly the issue. The issue is that I can't write to a file using the `sys.io.File` without the game just crashing (and no, it does not write to the file before, on, or after the crash). The path to the file is fine. My question was if there's a way to do what I'm trying to do a different way or not.

**Reply by Geokureli** [reply]
how do you know it's not crashing because of an incorrect file path? File and FileSystem are made to do the exact thing you're trying to do, rather than trying something else we should find out why this isn't working as it's supposed to.

For instance are you able to read the file using that file path? if you call
```hx
trace(FileSystem.absolutePath("assets/scripts/dialogue.txt"));
```
 what full path does it output? is it the one you expected?

**Reply by Geokureli** [reply]
I wrote this little test and it worked fine for me, on mac. give it a try and let me know if it works
```hx
var path = "assets/scripts/dialogue.txt";
trace('relative path: ' + path);
var content:String = null;
// check relative path
var exists = sys.FileSystem.exists(path);
if (exists)
{
	// get contents
	content = sys.io.File.getContent(path);
	trace('rel content: ' + content);
}
else // relative path didn't work, try absolute
{
	// convert to absolute
	path = sys.FileSystem.absolutePath(path);
	trace('abolute path: ' + path);
	//check exists
	exists = sys.FileSystem.exists(path);
	trace('abs exists: ' + exists);
	if (exists)
	{
		// get contents
		content = sys.io.File.getContent(path);
		trace('abs content: ' + content);
	}
}

if (content != null)
{
	// alter contents to show number of rewrites
	var SPLIT_ID = "number of writes: ";
	if (content.indexOf(SPLIT_ID) == -1)
	{
		trace("first time reading");
		// append number of writes 
		content += '\n ${SPLIT_ID}1';
	}
	else
	{
		// get the number of times written to
		var split = content.split(SPLIT_ID);
		var writes = Std.parseInt(split[1]);
		trace('written $writes times, iterating');
		// add 1
		content = split[0] + SPLIT_ID + (writes + 1);
	}
	
	// wrap in a try in case the file is write protected
	try
	{
		// write to the file
		sys.io.File.saveContent(path, content);
		trace("save successful");// not called if an error is thrown
	}
	catch (e)
	{
		// output the error
		trace('error saving to [$path], msg="$e"');
	}
}
```
I think the problem may be that you're using the linux path format: `folder/file.ext` rather than the windows-style `folder\\file.ext`. `absolutePath` should convert between the two to whatever format is necessary (openfl assets do this automatically I believe).

**Reply by itsamedood** [reply]
Well the game didn't crash, so that's progress, BUT it still didn't work :(

<img width="1178" alt="Screen Shot 2021-09-27 at 6 16 32 PM" src="https://user-images.githubusercontent.com/74739472/134992831-f649d7ab-d6b5-4463-8737-a27ceda6743f.png">

**Reply by Geokureli** [reply]
open finder, press Command + Shift + G and paste that absolute path, up to the last slash, without the filename. 

also in your project.xml do you have an asset's path specified? something like `<assets path="assets" />`? without that the compiler may not be copying the assets to the build folder

**Reply by itsamedood** [reply]
Ok so I figured out HALF the problem, and man do I feel stupid.

What you said got me thinking, so I checked `Project.xml`, and while I did have `<assets path="assets" />`, I had for whatever reason changed it to `<assets path="assets" embed="true" />`, which was causing issues.

Now it does find the file, and it's somewhat working. Here's the new problem:
```hx
try
{
    File.saveContent(FileSystem.absolutePath("assets/scripts/dialogue.txt"), "TEST");
    trace("Wrote successfully :)");
}
catch (e:Exception)
{
    trace("Error writing to file: [" + e + "]");
    trace("Wrote unsuccessfully :(");
}
```
As you can tell by this test, it should in theory write to dialogue.txt and inside should be just `TEST`, and say it wrote successfully, and do the opposite on error. It says it wrote successfully, however it doesn't actually save the content to the file. So close yet so far.

**Reply by Geokureli** [reply]
are you sure you're checking the `cc/export/macos/bin/CC.app/Contents/Resources/assets/scripts/dialogue.txt` version of the file rather than the `cc/assets/scripts/dialogue.txt` file? as I said before the compiler copies the file to the build location, so the original wouldn't show these changes

**Reply by itsamedood** [reply]
Ok so after checking the export it IS working, and I'm just stupid 😅

Thanks so much for your help man!
